domain/port/service/watcher.go

```diff
@@ -101,34 +104,61 @@ func (s *WatchableService) WatchOpenedPortsForApplication(ctx context.Context, a
 func (s *WatchableService) endpointToMachineMapper(
 	ctx context.Context, db database.TxnRunner, events []changestream.ChangeEvent,
 ) ([]changestream.ChangeEvent, error) {
+	if len(events) == 0 {
+		return nil, nil
+	}
+
+	var indexes []indexed
+	unique := make(map[unit.UUID]struct{})
+	for _, event := range events {
+		id, err := unit.ParseID(event.Changed())
+		if err != nil {
+			return nil, err
+		}
 
-	unitUUIDs, err := transform.SliceOrErr(events, func(e changestream.ChangeEvent) (unit.UUID, error) {
-		return unit.ParseID(e.Changed())
-	})
-	if err != nil {
-		return nil, err
+		indexes = append(indexes, indexed{event: event, id: id})
+		unique[id] = struct{}{}
 	}
 
-	machineNames, err := s.st.GetMachineNamesForUnits(ctx, unitUUIDs)
-	if err != nil {
-		return nil, err
+	// This is suboptimal, we need the unit.UUID in the result so we can map
+	// it back to the id.
+	reference := make(map[unit.UUID]coremachine.Name)
+	for id := range unique {
+		machineNames, err := s.st.GetMachineNamesForUnits(ctx, []unit.UUID{id})
+		if err != nil {
+			return nil, err
+		} else if len(machineNames) != 1 {
+			return nil, errors.Errorf("expected exactly one machine for unit %q, got %d", id, len(machineNames))
```

Comment:
There's no need to do this enforcement. Just leave it as it was. 

We can rely on the domain model for integrity.
- We only use this watcher in the non-CAAS firewaller.
- We know this can only be tripped by deployed units opening or closing ports.
- We know such units have a machine assignment.

The change could be for multiple units on a machine anyway, which would create duplicate notifications for the same name.

Reply:
I will make a task to fix this watcher in the future; this doesn't do what we want long term.

Reply:
Then just return an error if you got zero machines back for the units.